File: back/src/test/java/com/back/domain/wish/controller/ApiV1WishControllerTest.java
package com.back.domain.wish.controller;

import com.back.domain.book.entity.Book;
import com.back.domain.book.service.BookService;
import com.back.domain.member.entity.Member;
import com.back.domain.member.service.MemberService;
import com.back.domain.wish.controller.ApiV1WishController;
import com.back.domain.wish.entity.Wish;
import com.back.domain.wish.service.WishService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithUserDetails;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.context.WebApplicationContext;

import java.util.List;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@ActiveProfiles("test")
@SpringBootTest
@Transactional
public class ApiV1WishControllerTest {

    @Autowired
    private WebApplicationContext context;
    private MockMvc mvc;

    @Autowired
    private WishService wishService;
    @Autowired
    private MemberService memberService;
    @Autowired
    private BookService bookService;

    @BeforeEach
    void setup() {
        mvc = MockMvcBuilders.webAppContextSetup(context).build();
    }

    @Test
    @DisplayName("찜 목록 조회")
    @WithUserDetails("user1")
    void t1() throws Exception {

        Member actor = memberService.findByUsername("user1");

        List<Wish> wishes = wishService.findByMember(actor);
        int wishesSize = wishes.size();

        ResultActions resultActions = mvc
                .perform(
                        get("/api/v1/wishes/mine")
                                .contentType(MediaType.APPLICATION_JSON))
                .andDo(print());

        resultActions
                .andExpect(handler().handlerType(ApiV1WishController.class))
                .andExpect(handler().methodName("getWishes"))
                .andExpect(status().isOk());

        for (int i = 0; i < wishesSize; i++) {
            Book book = wishes.get(i).getBook();
            List<String> tags = bookService.getBookTags(book);
            resultActions
                    .andExpect(jsonPath("$.[%d].id".formatted(i)).value(book.getId()))
                    .andExpect(jsonPath("$.[%d].title".formatted(i)).value(book.getTitle()))
                    .andExpect(jsonPath("$.[%d].imgUrl".formatted(i)).value(book.getImgUrl()))
                    .andExpect(jsonPath("$.[%d].tags".formatted(i)).isArray())
                    .andExpect(jsonPath("$.[%d].averageRating".formatted(i)).value(book.getAverageRating()));

            for (int j = 0; j < tags.size(); j++) {

                resultActions
                        .andExpect(jsonPath("$[%d].tags[%d]".formatted(i, j)).value(tags.get(j)));

            }
        }
    }

    @Test
    @DisplayName("찜 추가")
    @WithUserDetails("user1")
    void t2() throws Exception {

        Long bookId = 2L;

        ResultActions resultActions = mvc
                .perform(
                        post("/api/v1/wishes/book/%d".formatted(bookId))
                                .contentType(MediaType.APPLICATION_JSON))
                .andDo(print());

        resultActions
                .andExpect(handler().handlerType(ApiV1WishController.class))
                .andExpect(handler().methodName("addWish"))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.resultCode").value("201-1"))
                .andExpect(jsonPath("$.message").value("찜 추가 성공"));
    }


    @Test
    @DisplayName("찜 삭제")
    @WithUserDetails("user1")
    void t3() throws Exception {

        Long bookId = 1L;

        ResultActions resultActions = mvc
                .perform(
                        delete("/api/v1/wishes/book/%d".formatted(bookId)))
                .andDo(print());

        resultActions
                .andExpect(handler().handlerType(ApiV1WishController.class))
                .andExpect(handler().methodName("deleteWish"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.resultCode").value("200-1"))
                .andExpect(jsonPath("$.message").value("찜 삭제 성공"));
    }
}

File: back/src/test/java/com/back/domain/wish/controller/ApiV1WishControllerTest.kt
package com.back.domain.wish.controller

import com.back.domain.book.service.BookService
import com.back.domain.member.service.MemberService
import com.back.domain.wish.entity.Wish
import com.back.domain.wish.service.WishService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithUserDetails
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.handler
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.context.WebApplicationContext

@ActiveProfiles("test")
@SpringBootTest
@Transactional
class ApiV1WishControllerTest {
    @Autowired
    private lateinit var context: WebApplicationContext
    private lateinit var mvc: MockMvc

    @Autowired
    private lateinit var wishService: WishService

    @Autowired
    private lateinit var memberService: MemberService

    @Autowired
    private lateinit var bookService: BookService

    @BeforeEach
    fun setup() {
        mvc = MockMvcBuilders.webAppContextSetup(context).build()
    }

    @Test
    @DisplayName("내 찜 다건 조회")
    @WithUserDetails("user1")
    @Throws(Exception::class)
    fun t1() {
        val actor = memberService.getByUsername("user1")

        val wishes: List<Wish> = wishService.getWishesByMember(actor)
        val wishesSize = wishes.size

        val resultActions = mvc
            .perform(
                MockMvcRequestBuilders.get("/api/v1/wishes/mine")
                    .contentType(MediaType.APPLICATION_JSON)
            )
            .andDo(print())

        resultActions
            .andExpect(handler().handlerType(ApiV1WishController::class.java))
            .andExpect(handler().methodName("getWishes"))
            .andExpect(status().isOk())

        for (i in 0..<wishesSize) {
            val book = wishes[i].book
            val tags: List<String> = bookService.getBookTags(book)
            resultActions
                .andExpect(jsonPath("$.[$i].id").value(book.id))
                .andExpect(jsonPath("$.[$i].title").value(book.title))
                .andExpect(jsonPath("$.[$i].imgUrl").value(book.imgUrl))
                .andExpect(jsonPath("$.[$i].tags").isArray())
                .andExpect(
                    jsonPath("$.[$i].averageRating").value(book.averageRating)
                )

            for (j in tags.indices) {
                resultActions
                    .andExpect(jsonPath("$[$i].tags[$j]").value(tags[j]))
            }
        }
    }

    @Test
    @DisplayName("찜 생성")
    @WithUserDetails("user1")
    @Throws(Exception::class)
    fun t2() {
        val bookId = 2L

        val resultActions = mvc
            .perform(
                post("/api/v1/wishes/book/$bookId")
                    .contentType(MediaType.APPLICATION_JSON)
            )
            .andDo(print())

        resultActions
            .andExpect(handler().handlerType(ApiV1WishController::class.java))
            .andExpect(handler().methodName("addWish"))
            .andExpect(status().isCreated())
            .andExpect(jsonPath("$.resultCode").value("201-1"))
            .andExpect(jsonPath("$.message").value("찜 생성을 성공했습니다."))
    }

    @Test
    @DisplayName("찜 삭제(구형)")
    @WithUserDetails("user1")
    @Throws(Exception::class)
    fun t3() {
        val bookId = 1L

        val resultActions = mvc
            .perform(
                delete("/api/v1/wishes/book/$bookId")
            )
            .andDo(print())

        resultActions
            .andExpect(handler().handlerType(ApiV1WishController::class.java))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.resultCode").value("200-1"))
            .andExpect(jsonPath("$.message").value("찜 삭제를 성공했습니다."))
    }

    @Test
    @DisplayName("찜 삭제")
    @WithUserDetails("user1")
    @Throws(Exception::class)
    fun t4() {
        val wishId = 1L

        val resultActions = mvc
            .perform(
                delete("/api/v1/wishes/$wishId")
            )
            .andDo(print())

        resultActions
            .andExpect(handler().handlerType(ApiV1WishController::class.java))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.resultCode").value("200-1"))
            .andExpect(jsonPath("$.message").value("찜 삭제를 성공했습니다."))
    }
}